redis/src/cluster_async/mod.rs

```diff
@@ -0,0 +1,1239 @@
+        T: IntoConnectionInfo + Send + 'a,
+    {
+        async move {
+            let connection_info = info.into_connection_info()?;
+            let client = crate::Client::open(connection_info)?;
+            client.get_multiplexed_tokio_connection().await
+        }
+        .boxed()
+    }
+}
+
+async fn connect_and_check<T, C>(info: T) -> RedisResult<C>
+where
+    T: IntoConnectionInfo + Send,
+    C: ConnectionLike + Connect + Send + 'static,
+{
+    let mut conn = C::connect(info).await?;
+    check_connection(&mut conn).await?;
+    Ok(conn)
+}
+
+async fn check_connection<C>(conn: &mut C) -> RedisResult<()>
+where
+    C: ConnectionLike + Send + 'static,
+{
+    let mut cmd = Cmd::new();
+    cmd.arg("PING");
+    cmd.query_async::<_, String>(conn).await?;
+    Ok(())
+}
+
+fn get_random_connection<'a, C>(
+    connections: &'a ConnectionMap<C>,
+    excludes: Option<&'a HashSet<String>>,
+) -> (String, ConnectionFuture<C>)
+where
+    C: Clone,
+{
+    debug_assert!(!connections.is_empty());
+
+    let mut rng = thread_rng();
+    let sample = match excludes {
+        Some(excludes) if excludes.len() < connections.len() => {
+            let target_keys = connections.keys().filter(|key| !excludes.contains(*key));
+            target_keys.choose(&mut rng)
+        }
+        _ => connections.keys().choose(&mut rng),
+    };
+
+    let addr = sample.expect("No targets to choose from");
+    (addr.to_string(), connections.get(addr).unwrap().clone())
+}
+
+fn slot_for_key(key: &[u8]) -> u16 {
+    let key = sub_key(&key);
+    State::<XMODEM>::calculate(&key) % SLOT_SIZE as u16
+}
+
+// If a key contains `{` and `}`, everything between the first occurence is the only thing that
+// determines the hash slot
+fn sub_key(key: &[u8]) -> &[u8] {
+    key.iter()
+        .position(|b| *b == b'{')
+        .and_then(|open| {
+            let after_open = open + 1;
+            key[after_open..]
+                .iter()
+                .position(|b| *b == b'}')
+                .and_then(|close_offset| {
+                    if close_offset != 0 {
+                        Some(&key[after_open..after_open + close_offset])
+                    } else {
+                        None
+                    }
+                })
+        })
+        .unwrap_or(key)
+}
+
+#[derive(Debug)]
+struct Slot {
+    start: u16,
+    end: u16,
+    master: String,
+    replicas: Vec<String>,
+}
+
+impl Slot {
+    pub fn start(&self) -> u16 {
+        self.start
+    }
+    pub fn end(&self) -> u16 {
+        self.end
+    }
+    pub fn master(&self) -> &str {
+        &self.master
+    }
+    #[allow(dead_code)]
+    pub fn replicas(&self) -> &Vec<String> {
+        &self.replicas
+    }
+}
+
+// Get slot data from connection.
+async fn get_slots<C>(addr: &str, connection: &mut C, use_tls: bool) -> RedisResult<Vec<Slot>>
+where
+    C: ConnectionLike,
+{
+    trace!("get_slots");
+    let mut cmd = Cmd::new();
+    cmd.arg("CLUSTER").arg("SLOTS");
```

Comment:
should use `CLUSTER SHARDS` on redis 7 - `CLUSTER SLOTS` is deprecated.
https://redis.io/commands/cluster-slots/

Reply:
Thanks, raised as separate issue.